Added gravatar url generation for contributors

GiacomoPignoni opened the pull request:
I have added the code to generate the gravatar url and I have added default avatar for all teams.

Progress on #72

rviscomi commented:
This is great thanks @GiacomoPignoni! I'll send you an invitation to the Developers team, assuming you're interested in contributing again :)

I'll give this a review soon.

rviscomi commented on src/config.py:
```diff
@@ -1,7 +1,11 @@
 import json
+import urllib.parse, hashlib
```
Expand each import to its own line in sorted order

rviscomi commented on src/config.py:
```diff
@@ -1,7 +1,11 @@
 
 config_json = {}
 
+default_avatar_folder_path = "/static/images/"
```
Per [PEP 8](https://www.python.org/dev/peps/pep-0008/#constants) style, use CAPS for constant naming.

rviscomi commented on src/config.py:
```diff
@@ -14,5 +18,15 @@ def update_config():
+      elif 'teams' not in contributor[1]:
+        contributor[1]['avatar_url'] = default_avatar_path
+      else:
+        gravatar_url = "https://www.gravatar.com/avatar/" + hashlib.md5(contributor[1]['gravatar'].lower().encode()).hexdigest() + ".jpg?"
```
Can you also add `www.gravatar.com` to our [img-src CSP policy](https://github.com/HTTPArchive/almanac.httparchive.org/blob/f3883291d56f6af359aefa8c60b5e0b86136a53d/src/csp.py#L20)?

rviscomi commented on src/config.py:
```diff
@@ -14,5 +18,15 @@ def update_config():
   with open('config/2019.json', 'r') as config_file:
     config_json['2019'] = json.load(config_file)
 
+    for contributor in config_json['2019']['contributors'].items():
```
To avoid all the `contributor[1]`s let's unpack the key and value
```suggestion
    for contributor_id, contributor in config_json['2019']['contributors'].items():
```

rviscomi commented on src/config.py:
```diff
@@ -14,5 +18,15 @@ def update_config():
     config_json['2019'] = json.load(config_file)
 
+    for contributor in config_json['2019']['contributors'].items():
+      if 'gravatar' not in contributor[1] and 'teams' in contributor[1]:
```
Contributors without a gravatar will have deterministically-generated default avatars from [this set](https://drive.google.com/file/d/1p0mh0hcHLoY19G7CHKVEkYHI_jdLJvmT/view?usp=sharing). We still need to cookie-cut each graphic out, but for now could you change this logic from team-based avatars to deterministically-generated? Let's just rename the team-based images to `images/avatars/1.jpg` and keep a constant in this file for the number of default avatars.

For example, hash the contributor ID and take the modulo to select a particular avatar.

Also, it might be cleaner to handle the case when contributors _do_ have a gravatar first, then in the `else` block select the default avatar.

rviscomi commented on src/templates/en/2019/contributors.html:
```diff
@@ -54,6 +54,7 @@
   </div>
   {% for id, contributor in config.contributors.items() %}
     <div class="contributor {{ ' '.join(contributor.teams) }}">
+      <img src="{{ contributor.avatar_url }}">
```
Is there any way to fall back to a default avatar if the gravatar URL 404s?

GiacomoPignoni commented:
I just added all requested changes
I also cut the avatars from the given image and I put them on a 200x200 px images
I added a default image for the gravatar URL 404s, this is a default image from gravatar. Actually in the "d" parameter I can put a custom url for a custom default image

rviscomi commented on src/config.py:
```diff
@@ -14,5 +19,12 @@ def update_config():
+        gravatar_url += urllib.parse.urlencode({'d': 'mp','s':str(AVATAR_SIZE)})
+        contributor['avatar_url'] = gravatar_url
+      else:
+        contributor['avatar_url'] = DEFAULT_AVATAR_FOLDER_PATH + str(hash(contributor_id) % AVATARS_NUMBER) + ".jpg"
```
IIUC this will produce avatar numbers in the range 0 - 14. The avatar filenames go from 1 - 15. Could you reconcile this either by adding 1 to the modulo output or renaming file 15 to 0?

rviscomi commented on src/config.py:
```diff
@@ -14,5 +19,12 @@ def update_config():
-
+    for contributor_id, contributor in config_json['2019']['contributors'].items():
+      if 'gravatar' in contributor:
+        gravatar_url = "https://www.gravatar.com/avatar/" + hashlib.md5(contributor['gravatar'].lower().encode()).hexdigest() + ".jpg?"
```
Nit: Mixed use of single and double quotes, prefer to use single quotes throughout.

rviscomi commented:
> I just added all requested changes
I also cut the avatars from the given image and I put them on a 200x200 px images
I added a default image for the gravatar URL 404s, this is a default image from gravatar. Actually in the "d" parameter I can put a custom url for a custom default image

Nice! Thanks again for working on this @GiacomoPignoni, really great stuff and much appreciated!

Just a couple of small comments then this is good to go.

GiacomoPignoni commented:
Done, sorry for these stupid mistakes
